prow/pjutil/abort.go

```diff
@@ -0,0 +1,117 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pjutil
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+// prowClient a minimalistic prow client required by the aborter
+type prowClient interface {
+	//ReplaceProwJob replaces the prow job with the given name
+	ReplaceProwJob(string, prowapi.ProwJob) (prowapi.ProwJob, error)
+}
+
+// ProwJobResourcesCleanup type for a callback function which it is expected to clean up
+// all k8s resources associated with the given prow job. It should do the best effort to
+// remove these resources, but if for any reason there is an error, it should only log a warning
+// message.
+type ProwJobResourcesCleanup func(pj prowapi.ProwJob)
+
+// ProwJobAborter is an interface for abstracting the prow job aborter behaviour
+type ProwJobAborter interface {
+	// TerminateOlderPresubmitJobs aborts all prow presubmit jobs from the given list that
+	// have a newer version, and call the callback on each aborted job
+	TerminateOlderPresubmitJobs(pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanup) error
+}
+
+// ProwJobAborter provides functionality to abort prow jobs
+type prowJobAborter struct {
+	pjc prowClient
+	log *logrus.Entry
+}
+
+// jobIndentifier keeps the information required to uniquely identify a prow job
+type jobIndentifier struct {
+	job          string
+	organization string
+	repository   string
+	pullRequest  int
+}
+
+// String returns the string representation of a prow job identifier
+func (i *jobIndentifier) String() string {
+	return fmt.Sprintf("%s %s/%s#%d", i.job, i.organization, i.repository, i.pullRequest)
+}
+
+//NewProwJobAborter creates a new ProwJobAborter
+func NewProwJobAborter(pjc prowClient, log *logrus.Entry) ProwJobAborter {
+	return &prowJobAborter{
+		log: log,
+		pjc: pjc,
+	}
+}
+
+// TerminateOlderPresubmitJobs aborts all presubmit jobs from the given list that have a newer version. It calls
+// the cleanup callback for each job before updating its status as aborted.
+func (a *prowJobAborter) TerminateOlderPresubmitJobs(pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanup) error {
+	dupes := map[jobIndentifier]int{}
+	for i, pj := range pjs {
+		if pj.Complete() || pj.Spec.Type != prowapi.PresubmitJob {
+			continue
+		}
+
+		ji := jobIndentifier{
+			job:          pj.Spec.Job,
+			organization: pj.Spec.Refs.Org,
+			repository:   pj.Spec.Refs.Repo,
+			pullRequest:  pj.Spec.Refs.Pulls[0].Number,
+		}
+		prev, ok := dupes[ji]
+		if !ok {
+			dupes[ji] = i
+			continue
+		}
+		cancelIndex := i
+		if (&pjs[prev].Status.StartTime).Before(&pj.Status.StartTime) {
+			cancelIndex = prev
+			dupes[ji] = i
+		}
+		toCancel := pjs[cancelIndex]
+
+		cleanup(toCancel)
```

Comment:
Can we add a TODO here that we should cancel the prowjob before cleaning up the pod and make these systems independent? Best to do this in a separate PR, and preserve plank functionality here.